### PR TITLE
yaoming's backend-python-test

### PR DESCRIPTION
**Pull request by yaomingyang2002**
This is yaoming yang's backend-python-test work

### CODE REVIEW

**Comment by yaomingyang2002** [comment]
use DataTables CDN and modified files to use pagination

File: alayatodo/views.py [hunk 1]
```diff
@@ -51,7 +50,6 @@ def todo(id):
     return render_template('todo.html', todo=todo)
 
 
-@app.route('/todo', methods=['GET'])
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Is this redundancy ?
Is it possible that this was simply registering multiple routes to the same endpoint ?

File: alayatodo/views.py [hunk 2]
```diff
@@ -63,6 +63,10 @@ def todos():
 def todos_POST():
     if not session.get('logged_in'):
         return redirect('/login')
+
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Would it be good to have frontend validation as well ?

File: alayatodo/views.py [hunk 3]
```diff
@@ -45,6 +45,15 @@ def todo(id):
     return render_template('todo.html', todo=todo)
 
 
+@app.route('/todo/<id>/json', methods=['GET'])
```
Comment (m-a-choquette) [inline alayatodo/views.py]: We use the content-type header instead of a separate route

File: alayatodo/views.py [hunk 4]
```diff
@@ -48,8 +48,9 @@ def todo(id):
 @app.route('/todo/<id>/json', methods=['GET'])
 def todo_json(id):
     if not session.get('logged_in'):
+        flash('Need to login to see a todo_json job!')
```
Comment (m-a-choquette) [inline alayatodo/views.py]: the results in an html response while the request is asking for json

File: alayatodo/views.py [hunk 5]
```diff
@@ -52,11 +52,11 @@ def todo_json(id):
         return redirect('/login')
     cur = g.db.execute("SELECT * FROM todos WHERE id =%s AND user_id = %s" % (id, session['user']['id']))
     todo = cur.fetchone()
-    print(todo)
     if todo:
-        return jsonify(id=todo[0], user_id=todo[1], Description=todo[2])
+        return jsonify(id=todo[0], user_id=todo[1], Description=todo[2], Completed=todo[3])
```
Comment (m-a-choquette) [inline alayatodo/views.py]: Could there be a better approach to handle the serialization of the entity ?

File: alayatodo/views.py [hunk 6]
```diff
@@ -92,3 +92,25 @@ def todo_delete(id):
     g.db.commit()
     flash('You have just DELETed a Todo job! Job_id = %s' % id)
     return redirect('/todo')
+
+
+@app.route('/todo/complete/<id>')
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This action supports `GET` requests ?

File: resources/todo.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from flask_sqlalchemy import SQLAlchemy
+db = SQLAlchemy()
+
+class Todo(db.Model):
+    __tablename__ = 'todos'
+
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(db.Integer, db.ForeignKey('users.id'))
+    description = db.Column(db.String)
+    completed = db.Column(db.String)
```
Comment (m-a-choquette) [inline resources/todo.py]: This should be a boolean

File: alayatodo/templates/todos.html [hunk 8]
```diff
@@ -14,23 +19,29 @@ <h1>Todo List:</h1>
                        {{ todo.description }}
                     </a>
                 </td>
+                <td>
+                    <a href="{{ url_for('todo_complete', id=todo.id) }}">
```
Comment (m-a-choquette) [inline alayatodo/templates/todos.html]: A `GET` request to complete a todo ?

File: resources/database.sql [hunk 9]
```diff
@@ -9,6 +9,7 @@ DROP TABLE IF EXISTS todos;
 CREATE TABLE todos (
   id INTEGER PRIMARY KEY,
   user_id INT(11) NOT NULL,
-  description VARCHAR(255),
+  description VARCHAR(255) NOT NULL,
+  completed VARCHAR (10) DEFAULT 'NO',
```
Comment (m-a-choquette) [inline resources/database.sql]: TASK 2: Write a database migration script in resources/
This should have been in a migration file to be able to update existing projects.

File: resources/todo.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+from flask_sqlalchemy import SQLAlchemy
```
Comment (m-a-choquette) [inline resources/todo.py]: this dependency (and others) is not in the requirements

File: alayatodo/views.py [hunk 11]
```diff
@@ -1,17 +1,20 @@
...
-    g,
-    redirect,
-    render_template,
-    request,
-    session
-    )
+from flask import (Flask, g, redirect, render_template, request, session, flash, jsonify)
+# from flask_sqlalchemy import SQLAlchemy
+# from resources.user import User
+# from resources.todo import Todo
```
Comment (m-a-choquette) [inline alayatodo/views.py]: why is everything commented ?

File: resources/todo.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+db = SQLAlchemy()
+
+class Todo(db.Model):
```
Comment (m-a-choquette) [inline resources/todo.py]: The models are created, but they're not used anywhere

**Comment by yaomingyang2002** [comment]
Hi, m-a-choquette,
Thank you for the review and I appreciate your time. I have answered and addressed those questions you asked.
I thought the test has a time-limitation and once I made a pull-request I cannot touch it anymore. So I did the test over the previous weekend, tested it locally and then did the pull-request for the test.
Let me know if you have more comment or suggestions. 
Thanks, 
yaoming

File: resources/database.sql (hunk 9 above)
Comment (yaomingyang2002) [reply resources/database.sql]: instead of using Flask-Migrate

**Pull request closed without merging**